ENH: Add support for merged binary labelmap segmentations

Sunderlandkyl opened the pull request:
This commit adds support for the visualization, editing, and reading/writing of merged binary labelmap segmentations.
The 2D segmentation display manager now creates a single actor for each unique vtkDataObject in the segmentation, which allows merged labelmaps to be rendered with a single actor, which allows fast reslicing of atlases containing hundreds segments.

Editing of segments will now occur on a single binary labelmap "layer" unless the user edits a segment with that will overlap and not overwrite another segment on the same layer. At this point, the editing segment is separated from the merged labelmap and is moved to a different "layer".
Segments layers can be collapsed into merged layers from the Segmentations module, or by calling vtkSegmentation::CollapseBinaryLabelmaps(bool safeMerge=true).

Saving segmentations in .seg.nrrd format includes several new attributes to accommodate merged labelmaps.
New .seg.nrrd attributes:
* SegmentX_Layer: List dimension in the NRRD file that contains the segment
* SegmentX_LabelValue: Voxel value used to represent the segment

_____

**Layer column shown in segment table (disabled by default):**
---

![image](https://user-images.githubusercontent.com/9222709/65971604-67105000-e436-11e9-9114-ec5cdc8ced25.png)

**Widget in Segmentations module for collapsing layers:**
---

![image](https://user-images.githubusercontent.com/9222709/65971656-81e2c480-e436-11e9-9083-2c6bce702553.png)

cpinter commented on Libs/vtkSegmentationCore/vtkSegmentation.cxx:
```diff
@@ -1658,3 +2169,250 @@ void vtkSegmentation::DeserializeConversionParameters(std::string conversionPara
+  vtkSegment* segment = this->GetSegment(segmentId);
+  if (!segment)
+    {
+    // TODO
```
Please don't leave TODOs in the code without at least some explanation. This will convert into noise even for you after a few weeks, and is not useful for anybody else.

Sunderlandkyl replied:
Sorry, not intentional, I'll go through and make sure I haven't missed any TODO.

cpinter replied:
No problem, this is what code reviews are for :)

cpinter commented on Libs/vtkSegmentationCore/vtkSegmentation.h:
```diff
@@ -304,7 +308,63 @@ class vtkSegmentationCore_EXPORT vtkSegmentation : public vtkObject
+  //BTX
+  bool GenerateMergedLabelmap(vtkOrientedImageData* mergedImageData, int extentComputationMode, vtkOrientedImageData* mergedLabelmapGeometry = nullptr,
+    const std::vector<std::string>& segmentIDs = std::vector<std::string>());
+  ///ETX
```
BTX/ETX is deprecated. I don't remember what's instead, because in every case I could safely remove it. Please try just removing it.
Other than that BTX has only 2 slashes while ETX has three, in case it needs to stay (remove third one from ETX)

cpinter commented on Libs/vtkSegmentationCore/vtkSegmentation.h:
```diff
@@ -304,7 +308,63 @@ class vtkSegmentationCore_EXPORT vtkSegmentation : public vtkObject
+  /// Returns true if the binary labelmap representation is merged
+  bool IsBinaryLabelmapMerged(std::string segmentID);
+
+  void GetMergedSegmentIDsForRepresentation(vtkSegment* segment, std::string representationName,
```
Please add documentation. It's far from trivial what is a merged segment ID list, what are shared segments, and what is a main segment. Thanks

Sunderlandkyl commented:
Sorry, that was from an older version. I changed it to: vtkSegmentation::CollapseBinaryLabelmaps(bool safeMerge=true).

Sunderlandkyl commented:
Comments have been addressed.

Sunderlandkyl commented:
@lassoan @jcfr @pieper
Any feedback or comments?

If possible, we would like to integrate this by the end of the week.

jcfr commented:
Do you observe a lag when user is drawing and it happens to overlap ?

pieper commented:
I don't have time for a close review, but the overall concept sounds great.  As an aside, the OHIF segmentation implementation is moving ahead and ended up with basically the same "layers of labelmaps" approach.  We've started discussions around adding this to DICOM SEG for efficiency (of object size, transfer rates, and encoding/decoding computation burden).  I'd say merge when you are ready and we can address any issues that arise.

fedorov commented:
Yes, I second Steve - it will be good to have this feature!

Sunderlandkyl commented:
@jcfr When the segments are all contained on the same layer, there should be no noticeable lag, even with hundreds of segments (ex Brain Atlas).
That being said, I just tested it, and there was some lag due to some optimizations that I mistakenly reverted in the segment editor undo/redo save state function. I'll fix that up and upload a video.

@federov @pieper Thanks, sounds good!

jcfr commented:
@Sunderlandkyl  Thanks for doing the additional testing and for working on a fix :pray:

fedorov commented:
> @Federov @pieper Thanks, sounds good!

`@Federov` should be @fedorov ;-)

Sunderlandkyl commented:
The other thing that we should nail down before integrating is the attribute names used for the .seg.nrrd. merged labelmap properties.

Currently they are:
- SegmentX_Layer
- SegmentX_LabelValue

Do these seem suitable, or does anyone have a different suggestion?

cpinter commented:
> the OHIF segmentation implementation is moving ahead and ended up with basically the same "layers of labelmaps" approach

@pieper Do you know if they still use PolySeg? Kyle just added many of these features in vtSegmentationCore (which is actually PolySeg - we still need to do the separation of the GitHub repos). I know they tried to use it for OHIF, but not sure if they ended up abandoning the idea.

pieper commented:
> @pieper Do you know if they still use PolySeg?

As far as I know, no, I don't think James is using PolySeg and OHIF is only supporting labelmap segmentation on 2D slices, but he's very keen on interoperability and getting the dicom SEG encoding right, so we should be able to do more head to head testing and compare notes on the implementations.

Sunderlandkyl commented:
Fix pushed.

Here is a video of the NAC brain atlas being edited as a merged labelmap:
https://www.dropbox.com/s/mfiateey65wphy8/MergedLabelmapDemo.mp4?dl=0

lassoan commented:
Can you upload a video with 3D model updates enabled?

Sunderlandkyl commented:
Here is a video with the 3D segment display enabled:
https://www.dropbox.com/s/hu3x6biue30kt2y/MergedLabelmapDemo2.mp4?dl=0

lassoan commented:
Thank you, this is really impressive!

jcfr commented:
> Thank you, this is really impressive!

Ditto. Really :heart: it, well done !

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/qSlicerSegmentEditorAbstractEffect.cxx:
```diff
@@ -348,27 +409,111 @@ void qSlicerSegmentEditorAbstractEffect::modifySelectedSegmentByLabelmap(vtkOrie
 
   // TODO: composite modifierLabelmap with threshold mask
```
is this TODO comment still valid?

lassoan reviewed (commented):
Wow, this has been a monumental effort. The implementation looks excellent, my comments mainly address clarity/simplicity of the API.

Sunderlandkyl commented on Base/Python/slicer/util.py:
```diff
@@ -1133,7 +1133,7 @@ def arrayFromGridTransformModified(gridTransformNode):
 def arrayFromSegment(segmentationNode, segmentId):
   """Return voxel array of a segment's binary labelmap representation as numpy array.
-  Voxels values are not copied.
+  Voxels values are copied.
```
Should I add a method to directly access the values of the merged labelmap as a numpy array?

lassoan replied:
Yes, that would be a good solution. Also describe how to get the label value for a segment (if it is complicated then you may add a convenient function for that, too).

Sunderlandkyl replied:
I added two functions arrayFromSegment and arrayFromOriginalSegment, the former of which is a copy of the thresholded labelmap, while the latter points to the original (potentially shared) representation.
arrayFromOriginalSegment doesn't seem to me to be a good name, but I'm not sure what else to call it.

Sunderlandkyl commented on Libs/vtkSegmentationCore/vtkSegment.h:
```diff
@@ -47,6 +47,12 @@ class vtkSegmentationCore_EXPORT vtkSegment : public vtkObject
+  enum
+  {
+    /// Invoked when content of the master representation in a segment is changed.
+    RepresentationObjectChanged = 72100,
```
Is it considered an abstraction violation for the vtkSegment to include vtkSegmentation?

lassoan replied:
There have been so many different kind of modifications that can be done in vtkSegment and so far a single modified event was sufficient. It would be nice if you could check how difficult/complicated would be to do without a special modified event for representation object change (and just use the simple Modified event).

Sunderlandkyl commented on Libs/vtkSegmentationCore/vtkSegmentation.h:
```diff
@@ -105,7 +107,9 @@ class vtkSegmentationCore_EXPORT vtkSegmentation : public vtkObject
     /// Invoked if segment IDs order is changed. Not called when a segment is added or removed.
-    SegmentsOrderModified
+    SegmentsOrderModified,
+    /// Invoked when content the data object of a representation is changed.
```
In this case, the event was meant to indicate that the underlying vtkDataObject has been changed to a different one. It does not include situations in which the segment representation has just been modified (which are covered by RepresentationModified and MasterRepresentationModified).

lassoan replied:
I see what's the difference now. For some reason, it is not common in VTK (or in Slicer, at the level of logic classes) to distringuish between these. I.e., if you modifiy points, it does not matter if you replaced the entire pointdata object or you modified point coordinates in  the same pointdata object. I know that in GUI classes this is an important distinction, and there are quite often different event for these, I'm just wondering why we don't usually need these at lower levels. Having two different events would complicate observations, as you would either need to add two observers (to detect either "change" or "modification") or you would have to fire modified event as well whenever you replaced the object.

Could you check how much it would complicate thing if you did not add the new event and just manage changes with the existing ones?

Sunderlandkyl commented on Libs/vtkSegmentationCore/vtkOrientedImageDataResample.h:
```diff
@@ -160,6 +164,23 @@ class vtkSegmentationCore_EXPORT vtkOrientedImageDataResample : public vtkObject
   static bool ApplyImageMask(vtkOrientedImageData* input, vtkOrientedImageData* mask, double fillValue, bool notMask = false);
 
+  /// Get the labelmap values contained within the mask
+  /// \param binaryLabelmap Input image to get values from
```
I made the change, but I've had a thought.

These functions are not necessarily labelmap-specific, even though they've been developed for (and probably will only be used) in that context.
In that case, would it be worth making them more generic?

lassoan replied:
What exactly do you have in mind? Could you give examples of methods that we have now and we would have if they were more generic?

Sunderlandkyl replied:
I have nothing in mind. It's just that most of the functions in vtkOrientedImageDataResample work on both labelmap and non-labelmap volumes.

lassoan replied:
I agree that it is good if we can find a name that can be interpreted more generally, but the name should be specific enough so that you can guess what it means.

lassoan replied:
OK, I see, the question is about calling it a labelValue or just value (can be int or float).

Since equality check for float values is very fragile, I would avoid using non-integer types for arguments that rely on equality check. Checking if larger/smaller (for example for thresholds) is safe.

Sunderlandkyl committed:
ENH: Add support for shared binary labelmap segmentations

This commit adds support for the visualization, editing, and reading/writing of shared binary labelmap segmentations.
The 2D segmentation display manager now creates a single actor for each unique vtkDataObject in the segmentation, which allows shared labelmaps to be rendered with a single actor, which allows fast reslicing of atlases containing hundreds segments.

Editing of segments will now occur on a single binary labelmap "layer" unless the user edits a segment with that will overlap and not overwrite another segment on the same layer. At this point, the editing segment is separated from the shared labelmap and is moved to a different "layer".
Segments layers can be collapsed into shared layers from the Segmentations module, or by calling vtkSegmentation::CollapseBinaryLabelmaps(bool forceToSingleLayer).

Saving segmentations in .seg.nrrd format includes several new attributes to accommodate shared labelmaps.
New .seg.nrrd attributes:
* SegmentX_Layer: The list dimension in the NRRD file that contains the segment
* SegmentX_LabelValue: The labelmap voxel value used to represent the segment

Sunderlandkyl commented:
I believe that all of the comments have been addressed.

Sunderlandkyl commented:
@lassoan Other changes integrated. This should be good to merge when you're ready.

lassoan commented:
Thank you, I'll merge tonight.

Which other pull requests should be merged along with this one? (I will search but maybe you have a list)

Sunderlandkyl commented:
Just two, SlicerRT and SlicerSegmentEditorExtraEffects.

https://github.com/SlicerRt/SlicerRT/issues/117
https://github.com/lassoan/SlicerSegmentEditorExtraEffects/issues/24

lassoan commented:
There is a test failure on Mac:

http://slicer.cdash.org/testDetails.php?test=9797305&build=1714731

Some time ago the test was passing (http://slicer.cdash.org/testDetails.php?test=9723836&build=1705938) but there were a number of changes since then, so the test failure may not be directly related to this merge. Anyway, @Sunderlandkyl it would be great if you could have a look.

Sunderlandkyl commented:
> Do we recompute the number of layers each time the segmentation is modified? Is it an expensive operation?

We update the number of layers every time vtkSegmentation::SegmentModified is invoked.
It's not an expensive operation, just counting the number of unique vtkDataObject.

lassoan commented:
> > Do we recompute the number of layers each time the segmentation is modified? Is it an expensive operation?
> 
> We update the number of layers every time vtkSegmentation::SegmentModified is invoked.
> It's not an expensive operation, just counting the number of unique vtkDataObject.

OK. Still, it may be better to only compute it if the section is opened. Especially if we add some more information, such as total number of segments (or later if we add more sophisticated GUI to move segments between layers). We do this for example in Models module GUI / Information section: we only compute the model volume and surface if that section is opened.
